src/modeling/xgboost_hyperparameter_tuning_cat.R

```diff
@@ -0,0 +1,387 @@
+
+  # Train model
+  cat("Starting site-stratified hyperparameter tuning...\n")
+
+  fold_models <- list()
+  fold_results <- list()
+  best_params <- list()
+
+  # Run through each train/val fold and hypertune
+  for (i in seq(1,n_folds)) {
+    cat(paste0("Tuning fold ", i, " of ", n_folds, "...\n"))
+
+    #setup performance dataframe
+    perf <- data.frame()
+
+    # Split data into training and validation sets based on indices above
+
+    train_data <- data%>%filter(fold_id != i)
+    val_data   <- data%>%filter(fold_id == i)
+
+    # set up weights (if none provided, default to 1)
+    if (is.null(weights)) {
+      w_train <- rep(1, nrow(train_data))
+      w_val   <- rep(1, nrow(val_data))
+    } else if (is.function(weights)) {
+      #set custom weights function
+      w_train <- weights(train_data)
+      w_val   <- weights(val_data)
+    } else {
+      stop("`weights` must be NULL or a function(data) -> numeric")
+    }
+
+    # set up data matrix for xgb
+    dtrain <- xgb.DMatrix(
+      data = as.matrix(train_data[, features]),
+      label = train_data[[target_col]],
+      weight = w_train # add weights
+    )
+
+    dval <- xgb.DMatrix(
+      data = as.matrix(val_data[, features]),
+      label = val_data[[target_col]],
+      weight = w_val # add weights
+    )
+
+    # Set train vs val for early stopping
+    watchlist <- list(train = dtrain, eval = dval)
+
+    #run through all hyper parameters and save to fold_models with name foldi_gridj
+    for (j in 1:nrow(tune_grid)) {
+
+      #setup parameters for tune
+      params <- list(
+        objective        = tune_grid$objective[j],
+        eval_metric      = tune_grid$eval_metric[j],
+        tree_method = "exact", #deterministic method for reproducibility
+        eta              = tune_grid$eta[j],
+        gamma            = tune_grid$gamma[j],
+        alpha            = tune_grid$alpha[j],
+        lambda           = tune_grid$lambda[j],
+        max_depth        = tune_grid$max_depth[j],
+        subsample        = tune_grid$subsample[j],
+        colsample_bytree = tune_grid$colsample_bytree[j],
+        min_child_weight  = tune_grid$min_child_weight[j],
+        num_class        = length(unique(data[[target_col]]))
+      )
+      #set seed for reproducibility
+      set.seed(123)
+
+      #train model with hyper parameters
+      model_ij <- xgb.train(
+        params = params,
+        data = dtrain,
+        nrounds = tune_grid$nrounds[j],
+        watchlist = watchlist,
+        #change early stopping rounds based on eta (smaller eta, larger rounds)
+        early_stopping_rounds = ifelse(tune_grid$eta[j] >= 0.1, 250,
+                                       ifelse(tune_grid$eta[j] >= 0.01, 500,
+                                              1000)),
+        print_every_n = 1000,
+        verbose = 0,
+        nthread = 1
+      )
+
+      pred_col <- paste0(target_col, "_guess")
+
+      # --- Validation Predictions ---
```

Comment:
```suggestion
      # ---- Validation Predictions ----
```

Suggestion: I like that you are differentiating these sections by what they do. Big function with lots of things going on, this will make editing later (if necessary) easier. I suggest using at least four dashes so that R's outliner will pick up these sections automatically and we can jump to them easily. I will make a code suggestion to show what I mean.

Reply:
Added!